src/wpublish/QueriesAndChunks.cc

```diff
@@ -489,105 +499,168 @@ QueriesAndChunks::ScanTableSumsMap QueriesAndChunks::_calcScanTableSums() {
-/// finished, which allows the scheduler to move on to another Task.
-/// If too many Tasks from the same user query are booted, all remaining tasks are moved to the snail
-/// scheduler in an attempt to keep a single user query from jamming up a scheduler.
 void QueriesAndChunks::_bootTask(QueryStatistics::Ptr const& uq, wbase::Task::Ptr const& task,
                                  wsched::SchedulerBase::Ptr const& sched) {
     LOGS(_log, LOG_LVL_INFO, "taking too long, booting from " << sched->getName());
+    auto bSched = _blendSched.lock();
+    if (bSched == nullptr) {
+        LOGS(_log, LOG_LVL_WARN, __func__ << " blendSched==null, can't boot task=" << task->getIdStr());
+        return;
+    }
+    if (bSched->isScanSnail(sched)) {
+        LOGS(_log, LOG_LVL_WARN, __func__ << " can't boot task, already on snail " << task->getIdStr());
+        return;
+    }
     sched->removeTask(task, true);
+    bool alreadyBooted = task->setBooted();
+    if (alreadyBooted) {
+        LOGS(_log, LOG_LVL_WARN, __func__ << task->getIdStr() << " was already booted");
+        return;
+    }
     uq->_tasksBooted += 1;
+    _bootedTaskTracker.addTask(task);
+    return;
+}
 
+void QueriesAndChunks::_bootUserQueries() {
     auto bSched = _blendSched.lock();
     if (bSched == nullptr) {
         LOGS(_log, LOG_LVL_WARN, "blendSched undefined, can't check user query");
         return;
     }
-    if (bSched->isScanSnail(sched)) {
-        // If it's already on the snail scan, it has already been booted from another scan.
-        if (uq->_tasksBooted > _maxTasksBooted + 1) {
-            LOGS(_log, LOG_LVL_WARN,
-                 "User Query taking excessive amount of time on snail scan and should be cancelled");
-            // TODO: Add code to send message back to czar to cancel this user query.
-        }
-    } else {
-        // Disabled as too aggressive and vulnerable to bad statistics. DM-11526
-        if (false && uq->_tasksBooted > _maxTasksBooted) {
-            LOGS(_log, LOG_LVL_INFO,
-                 "entire UserQuery booting from " << sched->getName() << " tasksBooted=" << uq->_tasksBooted
-                                                  << " maxTasksBooted=" << _maxTasksBooted);
-            uq->_queryBooted = true;
-            bSched->moveUserQueryToSnail(uq->queryId, sched);
-        }
-    }
-}
+    // Track the total number of Tasks that are running while booted
+    // and move the worst offenders to the snailScan. The worst offender is the UserQuery with
+    // the most `Task`s running that are taking too long.
+    // The following function only counts booted `Task`s that are still EXECUTING,
+    // and `countQIdVect` is order by worst offender to least worst offender.
+    auto [bootedTaskCount, countQIdVect] = _bootedTaskTracker.getTotalBootedTaskCount();
+    // Sum of Tasks in
+    int uqBootedTasksSum = 0;
+    int maxDarkQueries = 5 * _maxTasksBooted;
```

Comment:
Another magic number which hasn't been explained.

Reply:
That one is likely to stay and should be in a config file.